## Evaluation Metrics

---

Starter code for the evaluation mini-project

Start by copying your trained/tested POI identifier from that which you built in the validation mini-project

This is the second step toward building your POI identifier!    

Loading/formatting the data:

In [7]:
import pickle
import sys

sys.path.append("c:/pyprog/udamini/tools/")
from feature_format import featureFormat, targetFeatureSplit

data_dict = pickle.load(open("c:/pyprog/udamini/final_project/final_project_dataset.pkl", "r"))

Add more features to features_list!

In [8]:
features_list = ["poi", "salary"]

data = featureFormat(data_dict, features_list)
labels, features = targetFeatureSplit(data)

Your code goes here

In [12]:
from sklearn import model_selection

features_train, features_test, labels_train, labels_test = model_selection.train_test_split(
                                             features, labels, test_size=0.3, random_state=42)

from sklearn import tree
clf = tree.DecisionTreeClassifier()
clf.fit(features_train,labels_train)
clf.score(features_test,labels_test)

0.7241379310344828

How many POIs are in the test set for your POI identifier?

In [13]:
pred = clf.predict(features_test)
sum(pred)
print len([e for e in labels_test if e == 1.0])

4


How many people total are in your test set?

In [14]:
len(pred)

29

If your identifier predicted $0$ (not POI) for everyone in the test set, what would its **accuracy** be?

In [21]:
1.0 - 4.0/29

0.8620689655172413

**No** true positives!

Precision and recall can help illuminate your performance better

Use the **precision_score** and **recall_score** available in sklearn.metrics to compute those quantities

What’s the precision?

In [16]:
from sklearn.metrics import *

precision_score(labels_test, pred)

0.0

What’s the recall?

In [17]:
recall_score(labels_test, pred)

0.0

Here are some made-up predictions and true labels for a hypothetical test set; 

- fill in the following boxes to practice identifying true positives, false positives, true negatives, and false negatives

- let’s use the convention that “1” signifies a positive result, and “0” a negative

 - **6** true positives
 
 - **9** true negatives
 
 - **3** false positives
 
 - **2** false negatives

In [18]:
predictions = [0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1] 
true_labels = [0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0]

In [26]:
true_positive = 0
true_negative = 0
false_positive = 0
false_negative = 0

for p, t in zip(predictions, true_labels):
    if (p,t) == (1,1):
        true_positive += 1
    elif (p,t) == (0,0):
        true_negative += 1
    elif (p,t) == (1,0):
        false_positive += 1
    elif (p,t) == (0,1):
        false_negative += 1
    else:
        print("erro encontrado!")

print ('True Positives :{}, True Negatives :{}, False Positives :{}, False Negatives :{}'.format(true_positive, 
                                                                        true_negative, false_positive, false_negative))

True Positives :6, True Negatives :9, False Positives :3, False Negatives :2


What's the **precision** of this classifier?

In [19]:
precision_score(true_labels, predictions)

0.6666666666666666

What's the recall of this classifier?

In [20]:
recall_score(true_labels, predictions)

0.75

My **false positive** rate is high, which means that when a POI is present in the test data, I am good of flagging him or her


My identifier doesn’t have great **precision**, but it does have good **recall**. That means that, nearly every time a POI shows up in my test set, I am able to identify him or her. The cost of this is that I sometimes get some **false positives**, where non-POIs get flagged

*verificar conceito de F1 para esta questão! [aqui](https://en.wikipedia.org/wiki/F1_score)*


My identifier doesn’t have great **recall**, but it does have good **precision**. That means that whenever a POI gets flagged in my test set, I know with a lot of confidence that it’s very likely to be a real POI and not a false alarm. On the other hand, the price I pay for this is that I **sometimes miss real POIs**, since I’m effectively reluctant to pull the trigger on edge cases


My identifier has a really great **F1 score**

This is the best of both worlds. Both my false positive and false negative rates are **low**, which means that I can identify POI’s **reliably** and **accurately**. If my identifier finds a POI then the person is almost certainly a POI, and if the identifier does not flag someone, then they are almost certainly not a POI

*There’s usually a tradeoff between precision and recall - which one do you think is more important in your POI identifier? There’s no right or wrong answer, there are good arguments either way, but you should be able to interpret both metrics and articulate which one you find most important and why*

---

Go back to your code from the last lesson, where you built a simple first iteration of a POI identifier using a decision tree and one feature. Copy the POI identifier that you built into the skeleton code in evaluation/evaluate_poi_identifier.py 

Recall that at the end of that project, your identifier had an accuracy (on the test set) of 0.724. Not too bad, right? Let’s dig into your predictions a little more carefully

---

As you may now see, having imbalanced classes like we have in the Enron dataset (many more non-POIs than POIs) introduces some special challenges, namely that you can just guess the more common class label for every point, not a very insightful strategy, and still get pretty good accuracy!

Precision and recall can help illuminate your performance better. Use the **precision_score** and **recall_score** available in sklearn.metrics to compute those quantities.

---

*Note: you may see a message like UserWarning: The precision and recall are equal to zero for some labels. Just like the message says, there can be problems in computing other metrics (like the F1 score) when precision and/or recall are zero, and it wants to warn you when that happens*

Obviously this isn’t a very optimized machine learning strategy (we haven’t tried any algorithms besides the decision tree, or tuned any parameters, or done any feature selection), and now seeing the precision and recall should make that much more apparent than the accuracy did